### PR TITLE
add utility methods to set all debug options

### PR DESCRIPTION
**Pull request by reinsch82**
Signed-off-by: Reinhold Degenfellner <rdegenfellner@gradle.com>

### CODE REVIEW

**Comment by reinsch82** [comment]
I'm not sure which one is the better option:
- The `start...` method with multiple parameters or the methods for every single parameter?
- if the latter: should we turn on debugging if one of the options is set?

**Comment by reinsch82** [comment]
I'm not sure which one is the better option:
- The `start...` method with multiple parameters or the methods for every single parameter?
- if the latter: should we turn on debugging if one of the options is set?

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java [hunk 1]
```diff
@@ -189,8 +186,16 @@ public static void doNotPropagateSystemProperty(String name) {
     protected final GradleDistribution distribution;
     private GradleVersion gradleVersionOverride;
 
-    private boolean debug = Boolean.getBoolean(DEBUG_SYSPROP);
-    private boolean debugLauncher = Boolean.getBoolean(LAUNCHER_DEBUG_SYSPROP);
+    private DefaultJavaDebugOptions debug = createJavaDebugOptions(Boolean.getBoolean(DEBUG_SYSPROP));
```
Comment (wolfs) [inline subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java:189]: 🤔 I wouldn't use the production `DefaultJavaDebugOptions` in our test fixtures. I think it would be simpler to create a new class which doesn't use the Property API to configure our tests.

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java [hunk 2]
```diff
@@ -405,8 +410,8 @@ public GradleExecuter copyTo(GradleExecuter executer) {
             executer.withGradleVersionOverride(gradleVersionOverride);
         }
 
-        executer.startBuildProcessInDebugger(debug);
-        executer.startLauncherInDebugger(debugLauncher);
+        executer.startBuildProcessInDebugger(debug.getEnabled().get());
```
Comment (wolfs) [inline subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java:413]: ❌ Wouldn't you need to copy over the whole debug object to the new launcher?

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java [hunk 3]
```diff
@@ -1513,19 +1518,73 @@ public boolean isAllowExtraLogging() {
...
     @Override
     public GradleExecuter startBuildProcessInDebugger(boolean flag) {
-        debug = flag;
+        debug.getEnabled().set(flag);
         return this;
     }
 
+
+    @Override
+    public GradleExecuter startBuildProcessInDebugger(boolean flag, boolean server, boolean suspend, int port, String host) {
```
Comment (wolfs) [inline subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java:1527]:
```suggestion
    public GradleExecuter startBuildProcessInDebugger(boolean flag, Action<DebugOptions> debugOptions) {
```

🤔 Maybe an API like this would be better? Then you could configure the debug options in a lambda/Closure. For Groovy, we could also accept a `Closure` here with `@DelegatesTo` which would allow an even better API.

File: subprojects/launcher/src/main/java/org/gradle/launcher/daemon/configuration/DaemonBuildOptions.java [hunk 4]
```diff
@@ -30,35 +31,28 @@
...
-        options.add(new DebugSuspendOption());
-        options.add(new ApplyInstrumentationAgentOption());
-        options.add(new DaemonOption());
-        options.add(new ForegroundOption());
-        options.add(new StopOption());
-        options.add(new StatusOption());
-        options.add(new PriorityOption());
-        DaemonBuildOptions.options = Collections.unmodifiableList(options);
-    }
+    private static List<BuildOption<DaemonParameters>> options = ImmutableList.of(
```
Comment (wolfs) [inline subprojects/launcher/src/main/java/org/gradle/launcher/daemon/configuration/DaemonBuildOptions.java:39]: ❓   Is is a problem to use `ImmutableList` in launcher? It may be if our Guava version requires Java 8.

File: subprojects/wrapper/src/integTest/groovy/org/gradle/integtests/WrapperChecksumVerificationTest.groovy [hunk 5]
```diff
@@ -70,8 +70,12 @@ Expected checksum: 'bad'
...
 
         when:
-        def success = wrapperExecuter.run()
-
+        def success = wrapperExecuter
+            .withCommandLineGradleOpts("-Dorg.gradle.debug=true", "-Dorg.gradle.debug.server=false", "-Dorg.gradle.debug.host=localhost", "-Dorg.gradle.debug.port=5006")
+//            .withCommandLineGradleOpts(JvmOptions.getDebugArgument(false, false, "localhost:5005"))
+//            .withCommandLineGradleOpts("-agentlib:jdwp=transport=dt_socket,server=n,address=localhost:5005,suspend=y")
+//            .withBuildJvmOpts("-agentlib:jdwp=transport=dt_socket,server=n,address=localhost:5005,suspend=y")
+            .run()
```
Comment (wolfs) [inline subprojects/wrapper/src/integTest/groovy/org/gradle/integtests/WrapperChecksumVerificationTest.groovy:78]: ❌ I suppose this shouldn't be part of the PR.

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java (hunk 1 above)
Comment (reinsch82) [reply subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java:189]: I wasn't sure about that :).
I'll create a new class.

File: subprojects/launcher/src/main/java/org/gradle/launcher/daemon/configuration/DaemonBuildOptions.java (hunk 4 above)
Comment (reinsch82) [reply subprojects/launcher/src/main/java/org/gradle/launcher/daemon/configuration/DaemonBuildOptions.java:39]: as discussed:
SupportedBuildJvmIntegrationTest runs green, so we are fine here.

**Review by hythloda — APPROVED** [review]
Docs look good

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java [hunk 6]
```diff
@@ -1513,19 +1512,31 @@ public boolean isAllowExtraLogging() {
...
     }
 
     @Override
     public GradleExecuter startLauncherInDebugger(boolean flag) {
-        debugLauncher = flag;
+        debugLauncher.setEnabled(flag);
+        return this;
+    }
+
+    public GradleExecuter startLauncherInDebugger(Action<JavaDebugOptionsInternal> action) {
```
Comment (wolfs) [inline subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java:1532]: ❔ Did you check out how well this is usable from Groovy?

**Review by wolfs — COMMENTED** [review]
This looks mostly good now.

File: subprojects/docs/src/docs/userguide/reference/command_line_interface.adoc [hunk 7]
```diff
@@ -416,8 +416,8 @@ If set to `true` and debugging is enabled, Gradle will run the build with the so
 `-Dorg.gradle.debug.suspend=(true,false)`::
 When set to `true` and debugging is enabled, the JVM running Gradle will suspend until a debugger is attached. _Default is `true`._
 
-`-Dorg.gradle.daemon.debug=true`::
-Debug <<gradle_daemon.adoc#gradle_daemon, Gradle Daemon>> process.
+`-Dorg.gradle.daemon.debug=true` (deprecated)::
+Debug <<gradle_daemon.adoc#gradle_daemon, Gradle Daemon>> process. (duplicate of `-Dorg.gradle.daemon`)
```
Comment (eskatos) [inline subprojects/docs/src/docs/userguide/reference/command_line_interface.adoc:420]: :question: typo?
```suggestion
Debug <<gradle_daemon.adoc#gradle_daemon, Gradle Daemon>> process. (duplicate of `-Dorg.gradle.debug`)
```

File: subprojects/wrapper/src/integTest/groovy/org/gradle/integtests/WrapperChecksumVerificationTest.groovy [hunk 8]
```diff
@@ -84,7 +84,9 @@ Expected checksum: 'bad'
         file('gradle/wrapper/gradle-wrapper.properties') << "distributionSha256Sum=${Hashing.sha256().hashFile(distribution.binDistribution).toZeroPaddedString(Hashing.sha256().hexDigits)}"
 
         when:
-        def result = wrapperExecuter.withTasks("wrapper", "--gradle-version", "7.5").runWithFailure()
+        def result = wrapperExecuter.withTasks("wrapper", "--gradle-version", "7.5")
+            .withCommandLineGradleOpts(JvmOptions.getDebugArgument(false, false, "localhost:5006"), "-Dorg.gradle.debug=true", "-Dorg.gradle.debug.server=false")
+            .runWithFailure()
```
Comment (eskatos) [inline subprojects/wrapper/src/integTest/groovy/org/gradle/integtests/WrapperChecksumVerificationTest.groovy:89]: :question: is this change intended?

File: subprojects/docs/src/docs/userguide/reference/command_line_interface.adoc [hunk 9]
```diff
@@ -416,8 +416,8 @@ If set to `true` and debugging is enabled, Gradle will run the build with the so
 `-Dorg.gradle.debug.suspend=(true,false)`::
 When set to `true` and debugging is enabled, the JVM running Gradle will suspend until a debugger is attached. _Default is `true`._
 
-`-Dorg.gradle.daemon.debug=true`::
-Debug <<gradle_daemon.adoc#gradle_daemon, Gradle Daemon>> process.
+`-Dorg.gradle.daemon.debug=true` (deprecated)::
```
[the remark this answers is not in the document]
Comment (eskatos) [reply subprojects/docs/src/docs/userguide/reference/command_line_interface.adoc:419]: :x: I would either do all the deprecation work in another PR or in this one but not add `(deprecated)` here in this PR and the rest in another.

Deprecating this would require coverage that asserts that if used the user gets a proper deprecation warning.

**Review by eskatos — CHANGES_REQUESTED** [review]

File: subprojects/internal-integ-testing/src/test/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuterTest.groovy [hunk 10]
```diff
@@ -126,4 +128,32 @@ class AbstractGradleExecuterTest extends Specification {
         def e = thrown(IllegalArgumentException)
         e.message.contains("Builds cannot be started with the debugger enabled on CI")
     }
+
+    def "start build process in debugger options"() {
+        when:
+        executer.startBuildProcessInDebugger { it.server = false }
+
+        then:
+
```
Comment (wolfs) [inline subprojects/internal-integ-testing/src/test/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuterTest.groovy:137]: 💄 I'd remove the newline after `then:`.

**Review by wolfs — APPROVED** [review]

**Review by eskatos — APPROVED** [review]

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test and merge
